src/main/holochainManager.ts

```diff
@@ -250,21 +263,85 @@ export class HolochainManager {
 
   // TODO Add option to install happ without UI
   async installWebHapp(webHappPath: string, appId: string, networkSeed?: string) {
-    const uiTargetDir = this.fs.happUiDir(appId, this.holochainDataRoot);
-    console.log('uiTargetDir: ', uiTargetDir);
-    console.log('Installing app...');
-    const tempHappPath = await rustUtils.saveWebhapp(webHappPath, uiTargetDir);
-    console.log('Stored UI and got temp happ path: ', tempHappPath);
+    // Decode webhapp into .happ bytes and ui.zip bytes
+    // TODO Option to install headless app
+    const happAndUiBytes = await rustUtils.readAndDecodeWebhapp(webHappPath);
+
+    // write [sha256].happ to happs directory
+    const happHasher = crypto.createHash('sha256');
+    const happSha256 = happHasher.update(Buffer.from(happAndUiBytes.happBytes)).digest('hex');
+    const happFilePath = path.join(this.fs.happsDir(this.holochainDataRoot), `${happSha256}.happ`);
+    writeFile(happFilePath, Buffer.from(happAndUiBytes.happBytes));
+
+    // compute UI hash
+    const uiZipHasher = crypto.createHash('sha256');
+    const uiZipSha256 = uiZipHasher.update(Buffer.from(happAndUiBytes.uiBytes)).digest('hex');
+
+    // If same UI is not already stored, write ui bytes to temp file to compute hashes
+    // (randomly generated filename to prevent other processes from being able to modify
+    // the file before the hashes are computed)
+    const uiDir = path.join(this.fs.uisDir(this.holochainDataRoot), uiZipSha256);
```

Comment:
> is the content being unpacked and left on disk somewhere else so that we need to check the individual files?

Yes it is being unpacked into a directory later to be served to happ windows. Otherwise it would need to be unpacked each time an asset is served to the happ window which I think would be way too expensive. And the checking of hashes then happens on an asset by asset basis whenever it's being requested by a happ window.